File: 2025-09-25 - Deques/Exemplo_Python/TesteDeque.py
# ------------------------------------------------------------------------
# testeDeque.py
# Programa principal que demonstra todas as operações da Deque
# ------------------------------------------------------------------------
from Deque import Deque

def main():
   deque = Deque()

   print("=== Inserções iniciais ===")
   try:
      deque.inserirNoFim("TEC.0007", "Programação de Computadores", 120)
      print("Inserido TEC.0007 no fim.")
      deque.inserirNoFim("TEC.0027", "Estrutura de Dados Não-Lineares", 80)
      print("Inserido TEC.0027 no fim.")
      deque.inserirNoInicio("TEC.0142", "Programação para Redes", 120)
      print("Inserido TEC.0142 no início.")
      deque.inserirNoFim("TEC.0022", "Bancos de Dados", 80)
      print("Inserido TEC.0022 no fim.")
   except Exception as e:
      print("Erro:", e)

   print()
   deque.imprimirDeque()
   print("Tamanho:", deque.tamanho())
   print()

   print("=== Tentativa de duplicado TEC.0027 ===")
   try:
      deque.inserirNoFim("TEC.0027", "Duplicado", 90)
   except ValueError as e:
      print("Erro:", e)
   print()

   print("=== Busca TEC.0142 ===")
   try:
      n = deque.buscarPorCodigo("TEC.0142")
      print("Encontrado:", n)
   except KeyError as e:
      print("Erro:", e)
   print()

   print("=== Atualização TEC.0022 ===")
   try:
      deque.atualizarPorCodigo("TEC.0022", "Bancos de Dados Avançados", 90)
      print("Atualizado TEC.0022 com sucesso.")
   except KeyError as e:
      print("Erro:", e)
   print()

   deque.imprimirDeque()
   print("Tamanho:", deque.tamanho())
   print()

   print("=== Remoções início e fim ===")
   try:
      deque.removerDoInicio()
      print("Removido do início.")
      deque.removerDoFim()
      print("Removido do fim.")
   except IndexError as e:
      print("Erro:", e)
   print()

   deque.imprimirDeque()
   print("Tamanho:", deque.tamanho())
   print()

   print("=== Remoção por código TEC.0027 ===")
   try:
      deque.removerPorCodigo("TEC.0027")
      print("Removido TEC.0027 com sucesso.")
   except KeyError as e:
      print("Erro:", e)
   print()

   deque.imprimirDeque()
   print("Tamanho:", deque.tamanho())
   print()

   print("=== Esvaziamento total ===")
   while not deque.estaVazio():
      try:
         deque.removerDoInicio()
      except IndexError as e:
         print("Erro:", e)
   print("Deque esvaziada. Tamanho:", deque.tamanho())
   deque.imprimirDeque()

if __name__ == "__main__":
   main()

File: 2025-09-25 - Deques/Exemplo_C#/TesteDeque.cs
// ------------------------------------------------------------------------
// TesteDeque.cs
// Programa principal que demonstra todas as operações da Deque
// ------------------------------------------------------------------------
using System;
using System.Collections.Generic;

public class TesteDeque {
   public static void Main(string[] args) {
      Deque deque = new Deque();

      Console.WriteLine("=== Inserções iniciais ===");
      try {
         deque.InserirNoFim("TEC.0007", "Programação de Computadores", 120);
         Console.WriteLine("Inserido TEC.0007 no fim.");
         deque.InserirNoFim("TEC.0027", "Estrutura de Dados Não-Lineares", 80);
         Console.WriteLine("Inserido TEC.0027 no fim.");
         deque.InserirNoInicio("TEC.0142", "Programação para Redes", 120);
         Console.WriteLine("Inserido TEC.0142 no início.");
         deque.InserirNoFim("TEC.0022", "Bancos de Dados", 80);
         Console.WriteLine("Inserido TEC.0022 no fim.");
      } catch (Exception e) {
         Console.WriteLine("Erro: " + e.Message);
      }

      Console.WriteLine();
      deque.ImprimirDeque();
      Console.WriteLine("Tamanho: " + deque.Tamanho());
      Console.WriteLine();

      Console.WriteLine("=== Tentativa de duplicado TEC.0027 ===");
      try {
         deque.InserirNoFim("TEC.0027", "Duplicado", 90);
      } catch (ArgumentException e) {
         Console.WriteLine("Erro: " + e.Message);
      }
      Console.WriteLine();

      Console.WriteLine("=== Busca TEC.0142 ===");
      try {
         Node n = deque.BuscarPorCodigo("TEC.0142");
         Console.WriteLine("Encontrado: " + n.ToString());
      } catch (KeyNotFoundException e) {
         Console.WriteLine("Erro: " + e.Message);
      }
      Console.WriteLine();

      Console.WriteLine("=== Atualização TEC.0022 ===");
      try {
         deque.AtualizarPorCodigo("TEC.0022", "Bancos de Dados Avançados", 90);
         Console.WriteLine("Atualizado TEC.0022 com sucesso.");
      } catch (KeyNotFoundException e) {
         Console.WriteLine("Erro: " + e.Message);
      }
      Console.WriteLine();

      deque.ImprimirDeque();
      Console.WriteLine("Tamanho: " + deque.Tamanho());
      Console.WriteLine();

      Console.WriteLine("=== Remoções início e fim ===");
      try {
         deque.RemoverDoInicio();
         Console.WriteLine("Removido do início.");
         deque.RemoverDoFim();
         Console.WriteLine("Removido do fim.");
      } catch (InvalidOperationException e) {
         Console.WriteLine("Erro: " + e.Message);
      }
      Console.WriteLine();

      deque.ImprimirDeque();
      Console.WriteLine("Tamanho: " + deque.Tamanho());
      Console.WriteLine();

      Console.WriteLine("=== Remoção por código TEC.0027 ===");
      try {
         deque.RemoverPorCodigo("TEC.0027");
         Console.WriteLine("Removido TEC.0027 com sucesso.");
      } catch (KeyNotFoundException e) {
         Console.WriteLine("Erro: " + e.Message);
      }
      Console.WriteLine();

      deque.ImprimirDeque();
      Console.WriteLine("Tamanho: " + deque.Tamanho());
      Console.WriteLine();

      Console.WriteLine("=== Esvaziamento total ===");
      while (!deque.EstaVazio()) {
         try {
            deque.RemoverDoInicio();
         } catch (InvalidOperationException e) {
            Console.WriteLine("Erro: " + e.Message);
         }
      }
      Console.WriteLine("Deque esvaziada. Tamanho: " + deque.Tamanho());
      deque.ImprimirDeque();
   }
}
